lib/ukdebug/gdbstub.c

```diff
@@ -0,0 +1,291 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <uk/assert.h>
+#include <uk/essentials.h>
+#include <uk/arch/limits.h>
+
+#include <errno.h>
+#include <stddef.h>
+
+#define GDB_PACKET_RETRIES 5
+
+#define GDB_CHECK(expr) \
+	do { \
+		__ssz __r = (expr); \
+		if (__r < 0) \
+			return (int)__r; \
+	} while (0)
+
+int gdb_dbg_putc(char b __unused)
+{
+	return -ENOTSUP;
+}
+
+int gdb_dbg_getc(void)
+{
+	return -ENOTSUP;
+}
+
+static char gdb_checksum(const char *buf, __sz len)
+{
+	char c = 0;
+
+	while (len--)
+		c += *buf++;
+
+	return c;
+}
+
+static char *gdb_byte2hex(char *hex, __sz hex_len, char b)
+{
+	static const char map_byte2hex[] = "0123456789abcdef";
+
+	if (hex_len < 2)
+		return NULL;
+
+	*hex++ = map_byte2hex[(b & 0xf0) >> 4]; /* Encode high nibble */
+	*hex++ = map_byte2hex[(b & 0x0f) >> 0]; /* Encode low nibble */
+
+	return hex;
+}
+
+static __ssz gdb_mem2hex(char *hex, __sz hex_len, const char *mem, __sz mem_len)
+{
+	__sz l = mem_len;
+
+	if (hex_len < mem_len * 2)
+		return -ENOMEM;
+
+	while (l--)
+		hex = gdb_byte2hex(hex, 2, *mem++);
+
+	return mem_len * 2;
+}
+
+static int gdb_hex2int(char hex)
+{
+	if ((hex >= '0') && (hex <= '9'))
+		return hex - '0';
+	else if ((hex >= 'a') && (hex <= 'f'))
+		return hex - 'a' + 0xa;
+	else if ((hex >= 'A') && (hex <= 'F'))
+		return hex - 'A' + 0xa;
+
+	return -EINVAL;
+}
+
+static __ssz gdb_hex2mem(char *mem, __sz mem_len, const char *hex, __sz hex_len)
+{
+	__sz l = hex_len;
+	int i;
+
+	if (hex_len % 2 != 0)
+		return -EINVAL;
+
+	if (mem_len < hex_len / 2)
+		return -ENOMEM;
+
+	for (; l > 0; l -= 2, mem++) {
+		/* Decode high nibble */
+		i = gdb_hex2int(*hex++);
+		if (i < 0)
+			return i;
+
+		*mem = i << 4;
+
+		/* Decode low nibble */
+		i = gdb_hex2int(*hex++);
+		if (i < 0)
+			return i;
+
+		*mem |= i;
+	}
+
+	return hex_len / 2;
+}
+
+static __ssz gdb_send(const char *buf, __sz len)
+{
+	__sz l = len;
+
+	while (l--)
+		GDB_CHECK(gdb_dbg_putc(*buf++));
+
+	return len;
+}
+
+static __ssz gdb_recv(char *buf, __sz len)
+{
+	__sz l = len;
+	int r;
+
+	while (l--) {
+		r = gdb_dbg_getc();
+		if (r < 0)
+			return r;
+
+		*buf++ = (char)r;
+	}
+
+	return len;
+}
+
+static int gdb_send_ack(void)
+{
+	return gdb_dbg_putc('+');
+}
+
+static int gdb_send_nack(void)
+{
+	return gdb_dbg_putc('-');
+}
+
+static int gdb_recv_ack(void)
+{
+	int r = gdb_dbg_getc();
+
+	return (r < 0) ? r : ((char)r == '+');
```

Comment:
You are right, this is confusing. I changed it to be a function that returns a negative value on error and zero on success. It's easy to introduce stupid bugs if a function returns a boolean on success and a negative value on error. Because IIRC a negative non-zero value is interpreted as `true` in C and only `0` is interpreted as `false`.